[PowerToys Run] Add Suport for Commandline arguments in Program Plugin

royvou opened the pull request:
## Summary of the Pull Request

Add support for CommandLine arguments to the Program Plugin

## PR Checklist
* [x] Applies to #5724
* [x] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/PowerToys) and sign the CLA
* [x] Tests added/passed
* [ ] Requires documentation to be updated
* [ ] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan. Issue number where discussion took place: #xxx

## Info on Pull Request

Adds Support for CommandLine arguments to the Program Plugin.

## Validation Steps Performed
- Part 1
  - Ran an UWP Program with commandline arguments
  - Validate it ran as expected
- Part 2
  - Ran an Win32 Program with commandline arguments
  - Validate it ran as expected
- Part 3
  - Added Unit Tests :)

royvou commented:
The PR Is "Draft" because it needs some work to determine where the "Searchtext" should be resolved to the application / and where it would start passing the rest as arguments (E.g. Microsoft Edge is currently parsed as "Microsoft for application, Edge as arguments)).

Currently two options
"--"  seperates search text with argumentlist 
"-", "/" before a word (and not the first text, at least second or third)

htcfreek commented:
What about doing it like windows search: Accept parameters only when the user enter the exe file name.

inpute | result
------------ | -------------
firefox.exe -profile | firefox.exe -profile
firefox -profile | /no result/
Firefox | Mozilla Firefox

royvou commented:
@htcfreek i don't think this would be great, since it will not be possible to launch the programs via their description.

Currently i implemented in the following way.

- If the search query contains a "--", use everything behind it as the queryarguments
- If the search query (starting at term 2) contains a - or / it will parse that + rest as query arguments
- Otherwise parse everything to the query and keep arguments empty.

Doing my own testing now to see how i like the feature (you can too) and will add tests for the parsers later :)

htcfreek commented:
> The PR Is "Draft" because it needs some work to determine where the "Searchtext" should be resolved to the application / and where it would start passing the rest as arguments (E.g. Microsoft Edge is currently parsed as "Microsoft for application, Edge as arguments)).
> 
> Currently two options
> "--"  seperates search text with argumentlist 
> "-", "/" before a word (and not the first text, at least second or third)

What does happen if input is `edge.exe --inprivate` or `edge -- --inprivate`

royvou commented:
Added the tests for it as well :)

| Search | Command | arguments| Reason | 
--- | --- | --- | ---
| edge.exe --inprivate | edge.exe --inprivate   | No ' -- ' found and no single dash for arguments |
| edge -- --inprivate | edge | --inprivate | Found ' -- '

htcfreek commented:
> Added the tests for it as well :)
> 
> | Search | Command | arguments| Reason | 
> --- | --- | --- | ---
> | edge.exe --inprivate | edge.exe --inprivate   | | No ' -- ' found and no single dash for arguments |
> | edge -- --inprivate | edge | --inprivate | Found ' -- ' 

I think the first case should be also possible because for example chrome uses double minus for parameters.

Commit:
Implemented possibility to add commandline arguments in the Program Plugin

royvou commented:
Added support for the doubledash as well (And updated the regex to make it easier to add even more but don't think that will be needed). I thought the doubledash was more of the Linux style of commandline seperators but seems it's also used on Windows, so a good suggestion 👍

royvou commented on src/modules/launcher/Plugins/Microsoft.Plugin.Program/Programs/Win32Program.cs:
```diff
@@ -1,4 +1,4 @@
-// Copyright (c) Microsoft Corporation
```
This was a UTF8 files, but the other files in this project are UTF8 + BOM, seems like a sensible change to add?

royvou commented:
@htcfreek I think this all depends on how much / if any feedback we'll get on this feature 👍 but i assume will be close if everything goes great.
@crutkas were you able to try it out yet?

Yesterday i saw another merge conflict so i rebased it on top of that again.

htcfreek commented:
@crutkas 
Will we get this in for 0.21.0 or do we need more time for testing?

royvou commented:
Rebased it on the latest master branch to resolve merge conflict on the csproj.

htcfreek commented:
Can we merge this for 0.23?

@crutkas Have you tried this feature?

ryanbodrug-microsoft commented on src/modules/launcher/Plugins/Microsoft.Plugin.Program/Main.cs:
```diff
@@ -60,22 +68,35 @@ public void Save()
             }
 
-            return result.ToList();
+            return new List<Result>(0);
```
Thanks @royvou for this PR.  I apologize that I'm just taking a look now.   Doesn't this always return an empty list of programs?

royvou replied:
😭 Looks like in one of the merges I did not correctly merge/test it live. Unit tests were still passing (but those don't check the main program class.

Commit:
Merge remote-tracking branch 'microsoft/master' into feature/programcommandlinearguments

Commit:
Add missing return statement inc ommandArgumentParser loop

ryanbodrug-microsoft commented on src/modules/launcher/Plugins/Microsoft.Plugin.Program.UnitTests/ProgramArgumentParser/ProgramArgumentParserTests.cs:
```diff
@@ -0,0 +1,54 @@
+                new NoArgumentsArgumentParser(),
+           };
+
+            // Basis version oft he Quey parser which can be found at Wox.Core.Plugin.QueryBuilder  but didn't want to conenct this to that project
```
nit:typo

Commit:
Fix typos

ryanbodrug-microsoft commented on src/modules/launcher/Plugins/Microsoft.Plugin.Program.UnitTests/ProgramArgumentParser/ProgramArgumentParserTests.cs:
```diff
@@ -0,0 +1,54 @@
+        [TestCase("edge.exe -- --inprivate", "edge.exe", "--inprivate")]
+        [TestCase("edge.exe", "edge.exe", null)]
+        [TestCase("edge", "edge", null)]
+        [TestCase("cmd -c \"ping 1.1.1.1\"", "cmd", "-c \"ping 1.1.1.1\"")]
```
i don't think this works for cmd.  It might be better to supply a real world use case `cmd /c "ping 1.1.1.1"`

royvou replied:
Is a small improvement yes!

ryanbodrug-microsoft commented on src/modules/launcher/Plugins/Microsoft.Plugin.Program/Main.cs:
```diff
@@ -18,6 +19,13 @@ namespace Microsoft.Plugin.Program
     {
+        private static readonly IProgramArgumentParser[] _programArgumentParsers = new IProgramArgumentParser[]
+        {
+            new DoubleDashProgramArgumentParser(),
```
[discussion] Sort of a philosophical point, order matters here in that NoArgumentsArgumentParser is really a catch all for  when no parsers are able to handle the format.  A simple thing might be just to comment that 'NoArgumentsArgumentParser' needs to be last in the list to make it obvious to someone who wants to add a different parser.  If this was likely to be reused, we could consider implementing a `chain of responsibility` object that also allows for a default response if no 'node' in the chain can handle the given input.

ryanbodrug-microsoft replied:
For this PR I would just add a comment.  No need to overengineer at this point, but something to consider.

royvou replied:
Yes the order of the _programArgumentParsers array does matter a lot! It will use the first argumentparser which is compatible, so the 'most specific' parsers should be first (and the NoArgumentsParser should always be last).

Good suggestion to keep in mind 👍

Commit:
Fix Additional Typo

Commit:
Changed -c to /c to make it a valid cmd argument

Commit:
Added small comment about importance of order in _programArgumentParsers

ryanbodrug-microsoft reviewed (approved):
I really like this feature! Nicely done.

crutkas commented:
@royvou i was testing this due to an old issue i found, https://github.com/microsoft/PowerToys/issues/4497.

For CMD, this works but for terminal, this pattern does not.

htcfreek commented:
> @royvou i was testing this due to an old issue i found, https://github.com/microsoft/PowerToys/issues/4497.
> 
> For CMD, this works but for terminal, this pattern does not.

@crutkas 
What command have you entered?
I think it musst be written like `wt.exe -- ping github.com`

royvou commented:
@crutkas i would need to inspect but since neither use a - - or - / with argument I would expect it to be part of  the 'program' instead of the arguments.

royvou commented:
Current IProgramArgumentParser implementation assumes every keyword you type in (keywords are seperated by a space e.g. Microsoft Edge) would be part of the "Program" so `wt ping github.com` is determined to be the complete search string, which in turn will cause no arguments to be passed in :).

> does everything via shell so that's no problem (They Assume the first 'keyword' (You need to escape/quote it and it must be an exact match).

Not sure the best way to handle that cause, we can say use the '--' to seperate it, or for known programs (wt/pwsh/cmd come to mind) create a new argumentparser which special cases them :)
